example/poisson_gmg/poisson_driver.cpp

```diff
@@ -0,0 +1,566 @@
+    // level)
+    if (level < max_level) {
+      auto copy_over = post_smooth;
+      if (!do_FAS) {
+        copy_over = tl.AddTask(post_smooth, CopyData<u, res_err, true>, md);
+      } else {
+        auto calc_err = tl.AddTask(post_smooth, AddFieldsAndStore<u, u0, res_err, true>,
+                                   md, 1.0, -1.0);
+        copy_over = calc_err;
+      }
+      auto boundary =
+          AddBoundaryExchangeTasks<BoundaryType::gmg_same>(copy_over, tl, md, multilevel);
+      tl.AddTask(boundary, SendBoundBufs<BoundaryType::gmg_prolongate_send>, md);
+    } else {
+      AddBoundaryExchangeTasks<BoundaryType::gmg_same>(post_smooth, tl, md, multilevel);
+    }
+  }
+}
+
+TaskCollection PoissonDriver::MakeTaskCollectionMG(BlockList_t &blocks) {
+  using namespace parthenon;
+  using namespace poisson_package;
+  TaskCollection tc;
+  TaskID none(0);
+
+  auto pkg = pmesh->packages.Get("poisson_package");
+  auto max_iterations = pkg->Param<int>("max_iterations");
+
+  const int num_partitions = pmesh->DefaultNumPartitions();
+  int min_level = 0;
+  int max_level = pmesh->GetGMGMaxLevel();
+
+  {
+    TaskRegion &region = tc.AddRegion(num_partitions);
+    for (int i = 0; i < num_partitions; ++i) {
+      TaskList &tl = region[i];
+      auto &md = pmesh->mesh_data.GetOrAdd("base", i);
+      auto copy_exact = tl.AddTask(none, CopyData<exact, u>, md);
+      auto comm = AddBoundaryExchangeTasks<BoundaryType::any>(copy_exact, tl, md, true);
+      auto get_rhs = Axpy<u, u, rhs>(tl, comm, md, 1.0, 0.0, false, false);
+      auto zero_u = tl.AddTask(get_rhs, SetToZero<u>, md);
+    }
+  }
+
+  for (int ivcycle = 0; ivcycle < max_iterations; ++ivcycle) {
+    TaskRegion &region = tc.AddRegion(num_partitions * (max_level + 1));
+    for (int level = max_level; level >= min_level; --level) {
+      AddMultiGridTasksLevel(region, level, min_level, max_level, level == min_level);
+    }
+
+    int reg_dep_id = 0;
+    TaskRegion &region_res = tc.AddRegion(num_partitions);
+    for (int i = 0; i < num_partitions; ++i) {
+      TaskList &tl = region_res[i];
+      auto &md = pmesh->mesh_data.GetOrAdd("base", i);
+      auto calc_pointwise_res = Axpy<u, rhs, res_err>(tl, none, md, -1.0, 1.0, false);
+      auto get_res = DotProduct<res_err, res_err>(calc_pointwise_res, region_res, tl, i,
+                                                  reg_dep_id, &residual, md);
+      auto calc_err =
+          tl.AddTask(get_res, AddFieldsAndStore<u, exact, res_err>, md, 1.0, -1.0);
+      auto get_err = DotProduct<res_err, res_err>(calc_err, region_res, tl, i, reg_dep_id,
+                                                  &rhat0r, md);
+      if (i == 0) {
+        tl.AddTask(
+            get_err,
+            [&](PoissonDriver *driver) {
+              Real rms_res = std::sqrt(driver->residual.val / pmesh->GetTotalCells());
+              Real rms_err = std::sqrt(driver->rhat0r.val / pmesh->GetTotalCells());
```

Comment:
Is this normalization common rather than, say, a volume weighted one?

Reply:
Honestly, I didn't think too carefully about it. Nevertheless, I think this is the right thing to do since each zone corresponds to a row in the matrix equation and the residual we are reducing in BiCGSTAB for instance is just a vector with each element corresponding to a zone. To weight by zone volume, it might make sense to use a finite volume representation of the diffusion equation instead of a finite difference representation.